src/Kernel/Process.class.st

```diff
@@ -213,63 +213,46 @@ Process >> debugWithTitle: title [
-		let's find an inner context executing outerMost's argument block (tempAt: 1)"
-		(suspendedContext findContextSuchThat: [:ctx | 
-			ctx closure == outerMost unwindBlock]) ifNotNil: [:inner | 
-				"Let's finish the unfinished unwind context only (i.e. up to inner) and return here"
-				outerMost := (suspendedContext runUntilErrorOrReturnFrom: inner) first.
-				"Update the receiver's suspendedContext (the previous step reset its sender to nil);
-				return, if the execution stack reached its bottom (e.g. in case of non-local returns)."
-				(suspendedContext := outerMost sender) ifNil: [^self]]]. 
-
-	"Now all unwind blocks caught halfway through have been completed; 
-	let's execute the ones still pending. Note: #findNextUnwindContextUpTo: starts
-	searching from the receiver's sender but the receiver itself may be an unwind context."
-	ctxt := suspendedContext.
-	ctxt isUnwindContext ifFalse: [ctxt := ctxt findNextUnwindContextUpTo: nil].
-	[ctxt isNil] whileFalse: [
-		ctxt unwindComplete ifNil: [
-			ctxt unwindComplete: true.
-			unwindBlock := ctxt unwindBlock.
-			"Create a context for the unwind block and execute it on the unwind block's stack. 
-			Note: using #value instead of #runUntilErrorOrReturnFrom: would lead to executing 
-			the unwind on the wrong stack preventing the correct execution of non-local returns."
-			suspendedContext := unwindBlock asContextWithSender: ctxt.
-			suspendedContext runUntilErrorOrReturnFrom: suspendedContext].
-		ctxt := ctxt findNextUnwindContextUpTo: nil].
-
-	"Mark the context as terminated"
-	suspendedContext terminateTo: nil.
-	suspendedContext setSender: nil receiver: self method: (Process>>#endProcess) arguments: {}
+	It assumes that self is not the active process and the termination is requested from another process"	
+		
+	| ctxt oldList outerMost top newTop |
+
+	[ ] ensure: [ "Execute termination as an ensure block to ensure it completes even if terminated by another process before the termination is finished; see testTerminateInTerminate."
+		self isTerminating ifTrue: [ ProcessAlreadyTerminating signal. ^self ].
+		terminating := true.
+		"Always suspend the process first so it doesn't accidentally get woken up"
+		oldList := self suspend.
+		suspendedContext ifNil: [^self].
+		"Figure out if we are terminating the process while waiting inside special object like in Semaphore>>critical:. So if waiting object is interesting on this situation we will ask it to handle it. In case of Semaphore>>critical, Semaphore will pop the suspendedContext so that we leave the ensure: block inside Semaphore>>critical: without signaling the semaphore.
+		This methods allow to not be restricted only on Semaphore case."
+		suspendedContext := oldList handleProcessTerminationOfWaitingContext: suspendedContext.
+
+		top := suspendedContext.
+		suspendedContext := nil.  "Disable terminating process while running its stack in active process below.
```

Comment:
I am not sure about the idea of nil'ed suspendedContext. When debugger shows ZeroDivide in my example I can check the process in the playground: 
- it is terminated: p isTerminated "==> true"
- p suspendedContext "==> is nil"

This state is lying because the process is not yet terminated: it is in the middle of termination.  Why this nil is required? We have #terminating flag which protects from multiple termination.

Reply:
Yes, #terminating could be enough... I don't think nil is "required" - I didn't use it for anything; it just seemed safer just nil it. suspendedContext of a terminating process has no longer any meaning during termination (except the last step) because it's not used during termination. But you're right it implicates #isTerminated value... I'll think about this. However, if there's no practical use for suspendedContext or isTerminated showing false during termination - I'd say it's more a philosophical question :)

Reply:
> Yes, #terminating could be enough... I don't think nil is "required" - I didn't use it for anything; it just seemed safer just nil it. 

I prefer to keep the system simple. Any line of code raises the question "why it is here?". It's better to remove redundant code. Otherwise you should describe the need for this code with a test.

>> suspendedContext of a terminating process has no longer any meaning during termination (except the last step) because it's not used during termination. 

But it is used during termination. All unwinds are started from the suspendedContext. I think the unwind logic can better reflect the unwind steps: it can assign #newTop to #suspendedContext on each iteration. It will make the "outside termination" to be consistent with the "inside termination" (from the process itself).

>> However, if there's no practical use for suspendedContext or isTerminated showing false during termination - I'd say it's more a philosophical question :)

Having the system in the consistent state is very important. Especially when we implement such a very corner cases. It's always hard to debug and when in one place the process is shown as dead and in another place it is not dead that is a nightmare to figure things out.

Reply:
Hi Denis,
yes, you have a point indeed and I hear your arguments! My objective now is to make sure it works flawlessly (and I have tested it quite thoroughly) before any further polishing. But I'm working on your comments, thanks again!

Reply:
Hi, thanks for these great points!
> > Yes, #terminating could be enough... I don't think nil is "required" - I didn't use it for anything; it just seemed safer just nil it.
> 
> I prefer to keep the system simple. Any line of code raises the question "why it is here?". It's better to remove redundant code. Otherwise you should describe the need for this code with a test.
> 
Oh, I know why - if the suspendedContext is not nil the process being terminated can be resumed (e.g. by a higher priority process or thanks to preemption) and that would interfere with its termination in progress. Setting suspendedContext to nil prevents such scenario (see Process >> resume). In case this makes a good sense, I will update the comment; even I forgot the reasoning :D 

> > > suspendedContext of a terminating process has no longer any meaning during termination (except the last step) because it's not used during termination.
> 
> But it is used during termination. All unwinds are started from the suspendedContext. I think the unwind logic can better reflect the unwind steps: it can assign #newTop to #suspendedContext on each iteration. It will make the "outside termination" to be consistent with the "inside termination" (from the process itself).
> 
yes, right; I meant it has no reasonable meaning beyond the point the real unwinds start (i.e. after top := suspendedContext). The point is even if I set suspendedContext to newTop each time, its consistency is lost when the active process jumps onto the terminating process's stack and starts executing it's code. I remember when I first drafted this new approach I used suspendedContext the way you suggest but it kept failing for some reason so I gave up and left it as nil and used a variable instead (I'm not saying it's wrong; I just didn't manage to make it work that way).

> > > However, if there's no practical use for suspendedContext or isTerminated showing false during termination - I'd say it's more a philosophical question :)
> 
> Having the system in the consistent state is very important. Especially when we implement such a very corner cases. It's always hard to debug and when in one place the process is shown as dead and in another place it is not dead that is a nightmare to figure things out.
>
yes, I couldn't agree more; consistency is absolutely key; however in this case I don't know how to make isTerminated showing false during termination when I need suspendedContext = nil without further complicating isTerminated; actually maybe we could make isTerminated false if isTerminating = true -?? That should fix the consistency issue, what do you think? I'm not sure if all tests would agree with this change right away but it could be dealt with independently of this PR. Other option could be to change Process >> resume to take isTerminating into account and then we could remove nilling suspendedContext and leave isTerminated as is -?? (still independently of this PR)

Reply:
> Oh, I know why - if the suspendedContext is not nil the process being terminated can be resumed (e.g. by a higher priority process or thanks to preemption) and that would interfere with its termination in progress. Setting suspendedContext to nil prevents such scenario (see Process >> resume). In case this makes a good sense, I will update the comment; even I forgot the reasoning :D

Right, I see that #resume method contains nil check for suspendedContext.
I think for the real solution we should explicitly model the process status - possibly replace #isTerminating flag with #status variable (choose a better name) and reify the ProcessStatus hierarchy. It will allow to decouple the life cycle state of a process from its execution state like suspendedContext, pc, endProcess method.

Anyway it's for future work. PR is already a great improvement.

Reply:
Hi @dionisiydk , you wrote
> I think the unwind logic can better reflect the unwind steps: it can assign #newTop to #suspendedContext on each iteration.

FYI, you're right, it works ok and makes isTerminated answer correctly; the only issue is you're not protected against accidental resumption of the terminating process.